File: Javascript stack/case1rawplus.js
//		Start runtime timer and declare all global variables for the performance check.
const {performance} = require('perf_hooks');
var scriptStart = performance.now();
var scriptEnd;
var productLapStart;
var productLapEnd;
var printStart;
var printEnd;
var productCounter;
//		Import dependencies.
const puppeteer = require('puppeteer');
//		Declare the scraping method.
let scrape = async() => {
	//		Declare the URL and appropriate arguments.
	let fish = 'https://www.fcomputer.dk/computer/baerbar';
	let chips = '?eachPage=99999';
	//		Set some options for our driver, like a fake user-agent string and the interface mode.
	let userAgentString = 'Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:59.0) Gecko/20100101 Firefox/59.0';
	const driver = await puppeteer.launch({headless: true});
	const page = await driver.newPage();
	await page.setUserAgent(userAgentString);
	//		Pass the URL to our driver and tell it to wait max 2 minutes until there are no connections within half a second.
	await page.goto(fish+chips, {
		waitUntil: 'networkidle0',
		timeout: 120000,
	});
	//		Declare the timer and the method for data gathering within the page's context evaluation.
	productLapStart = performance.now();
	const soup = await page.evaluate(()=>{
		//		Declare an array to store the data and the products path to scrape.
		let data = [];
		let products = document.querySelectorAll('div.related-product-list-inner-wrapper');
		//		Declare the method to check existence of DOM elements and return a conditional value.
		function checkExistence(product, path, fallback){
			check = product.querySelector(path);
			//		Store inner text data as a string and specify a default value.
			check != null ? check = check.innerText : check = fallback;
			return check;
		};
		//		Declare a loop through the DOM tree to check existence and assign an appropriate value for each product attribute.
		for (var product of products) {
			let productName = checkExistence(product, 'span.product-header>div:nth-of-type(1)', '---');
			let productID = checkExistence(product, 'div.bottom-name-wrapper>span:nth-of-type(1)', '---');
			let productPrice = checkExistence(product, 'span.overview-product-price', 0);
			let productDiscount = checkExistence(product, 'div.inner-price-comparison-wrapper>span:nth-of-type(2)', 0);
			let productInCph = checkExistence(product, 'div.store-stock-wrapper:nth-of-type(1)>span.stock-title', 'N/A');
			let productInAar = checkExistence(product, 'div.store-stock-wrapper:nth-of-type(3)>span.stock-title', 'N/A');
			let productInWeb = checkExistence(product, 'div.store-stock-wrapper:nth-of-type(5)>span.stock-title', 'N/A');
			//		Pass the results to the general data array as a dictionary with key/value pairs.
			data.push({
				productName: productName,
				productID: productID,
				productPrice: productPrice,
				productDiscount: productDiscount,
				productInCph: productInCph,
				productInAar: productInAar,
				productInWeb: productInWeb
			});
		}
		//		Return the data evaluated within the page's context.
		return data;
	})
	//		Close the interface and return all the resulted data and variables.
	driver.close();
	return soup;
}
//		Initiate the scraping method and pass all the resulted data and variables for further manipulation.
scrape().then((value)=>{
	//		Declare a method to change the default value in the case of two specific conditions.
	function filterUnit(unit) {
		var unit = unit;
		if(unit != 'N/A') {
			//		Either if it's more than five, or any other amount.
			unit.match(/\+/g) ? unit = '+5' : unit = unit.replace(/\D/g, '')
		}
		return unit;
	}
	//		Initiate a loop through the data for manipulations.
	for (var i = 0; i < value.length; i++) {
		//		Remove irrelevant data from the product ID
		value[i].productID = value[i].productID.replace('Produkt nr.: ', '');
		//		Strip the digits and parse the result as an integer if the recieved value for the price tags is not default.
		if(value[i].productPrice != 0) value[i].productPrice = parseInt(value[i].productPrice.replace(/\D/g, ''));
		if(value[i].productDiscount != 0) value[i].productDiscount = parseInt(value[i].productDiscount.replace(/\D/g, ''));
		//		Change the availability status with the previously declared function.
		value[i].productInCph = filterUnit(value[i].productInCph);
		value[i].productInAar = filterUnit(value[i].productInAar);
		value[i].productInWeb = filterUnit(value[i].productInWeb);
	}
	//		Stop the data and runtime timers.
	productLapEnd = performance.now();
	scriptEnd = performance.now();
	//		Start the output timer and loop through the data to log it in the console.
	printStart = performance.now();
	for (var i = 0; i < value.length; i++) {
		//		Make sure that you don't mix up string and integer types.
		console.log("\r\n              DATA FOR ITEM #"+(i+1).toString());
		console.log("Name:                   "+value[i].productName);
		console.log("ID:                     "+value[i].productID);
		console.log("Price:                  "+value[i].productPrice.toString());
		console.log("Discount:               "+value[i].productDiscount.toString());
		console.log("Amount in Copenhagen:   "+value[i].productInCph);
		console.log("Amount in Copenhagen:   "+value[i].productInAar);
		console.log("Amount in the webshop:  "+value[i].productInWeb);
	}
	//		Stop the output timer and prepare the data amount.
	printEnd = performance.now();
	productCounter = value.length;
	//		Log the gathered performance data.
}).then(()=>{
	console.log("\r\nGathered 7 data attributes about "+productCounter.toString()+" products")
	console.log((productLapEnd-productLapStart)+" milliseconds to scrape and format the data")
	console.log((scriptEnd-scriptStart)+" milliseconds for script execution")
	console.log((printEnd-printStart)+" bonus milliseconds for logging the processed data")
});

//		The output should look like this:
//////// Gathered 7 data attributes about ___ products
//////// ___.___ milliseconds to scrape and format the data
//////// ___.___ milliseconds for script execution
//////// ___.___ bonus milliseconds for logging the processed data

File: Python stack/case1rawplus.py
# 		Start the runtime timer.
import time
scriptStart = time.time()
# 		Import dependencies.
from selenium import webdriver
from selenium.webdriver.firefox.options import Options
import bs4 as bs # BeautifulSoup.
import re # default Python library for regular expressions.
#		Set some options for our driver, like a fake user-agent string and the interface mode.
options = Options()
profile = webdriver.FirefoxProfile()
profile.set_preference("general.useragent.override", "Mozilla/5.0 (X11; Linux x86_64; rv:17.0) Gecko/20121202 Firefox/17.0 Iceweasel/17.0.1")
options.profile = profile
options.set_headless(headless=True)
driver = webdriver.Firefox(firefox_options = options)
#		Set the URL and appropriate arguments.
fish = "https://www.fcomputer.dk/computer/baerbar"
chips = "?eachPage=99999"
#		Pass the URL to the driver and get the source code.
driver.get(fish+chips)
pageSource = driver.page_source
#		Parse the recieved HTML, make sure to explicitly specify the parser to avoid possible errors.
soup = bs.BeautifulSoup(pageSource, "html.parser")
#		Declare your methods outside the scraping loop for optimize performance.
def unitCheck(product, location):
	#		This method should set a string that should have a value depending on one of three conditions.
	unit = product.find_all("div", class_="store-stock-wrapper")[location]
	if unit:
		#		If there's more than five units
		if re.search(r"[+]", unit.text) != None: unit = "5+"
		#		If there's five or less
		else: unit = re.sub('\D+', '', unit.text)
	#		Or if product count is not shown for any reason.
	else: unit = "N/A"
	return unit
#		Declare an array to store the data and start the timer for data gathering.
data = []
productLapStart = time.time()
#		Initialize a loop through the product containing elements.
for product in soup.find_all("div", class_="related-product-list-inner-wrapper"):
	#		Store the name and ID of the product as strings, cutting off the irrelevant data.
	productName = product.find("span", class_="product-header").find("div").text
	productID = product.find("div", class_="bottom-name-wrapper").find("span").text.replace("Produkt nr.: ", "")
	#		Check if price tags exist, strip the digits and parse the result as an integer.
	productPrice = product.find("span", class_="overview-product-price")
	if productPrice: productPrice = int(re.sub('\D+', '', productPrice.text))
	else: productPrice = None;
	productDiscount = product.find("div", class_="inner-price-comparison-wrapper")
	if productDiscount: productDiscount = int(re.sub('\D+', '', productDiscount.find_all("span")[1].text))
	else: productDiscount = None
	#		Store the availability status with the previously declared function.
	productInCph = unitCheck(product, 0)
	productInAar = unitCheck(product, 1)
	productInWeb = unitCheck(product, 2)
	#		Declare a dictionary data type for each product and pass key/value pairs to it.
	values = {}
	values.update({'productName':productName})
	values.update({'productID':productID})
	values.update({'productPrice':productPrice})
	values.update({'productDiscount':productDiscount})
	values.update({'productInCph':productInCph})
	values.update({'productInAar':productInAar})
	values.update({'productInWeb':productInWeb})
	#		End the loop with passing the data to the general array.
	data.append(values)
#		Stop the data timer and close the interface.
ProductLapEnd = time.time()
driver.quit()
#		Stop the runtime timer.
scriptEnd = time.time()
#		Start the output timer and loop through the data to log it in the console.
printStart = time.time()
productCounter = 1;
for item in data:
	#		Make sure that you don't mix up string and integer types.
	print("\r\n              DATA FOR ITEM #"+str(productCounter))
	print("Name:                   "+item['productName']),
	print("ID:                     "+item['productID']),
	print("Price:                  "+str(item['productPrice'])),
	print("Discount:               "+str(item['productDiscount'])),
	print("Amount in Copenhagen:   "+item['productInCph']),
	print("Amount in Copenhagen:   "+item['productInAar']),
	print("Amount in the webshop:  "+item['productInWeb']),
	productCounter+=1
#		Stop output timer and log the gathered performance data.
printEnd = time.time()
print("\r\nGathered 7 data attributes about "+str(productCounter-1)+" products")
print("{0:0.3f} seconds to scrape and format the data".format(ProductLapEnd-productLapStart))
print("{0:0.3f} seconds for script execution".format(scriptEnd-scriptStart))
print("{0:0.3f} bonus seconds for logging the processed data".format(printEnd-printStart))

#		The output should look like this:
#### Gathered 7 data attributes about ___ products
#### ___.___ seconds to scrape and format the data
#### ___.___ seconds for script execution
#### ___.___ bonus seconds for logging the processed data